File: 2551-apply-operations-to-an-array/2551-apply-operations-to-an-array.java
class Solution {
    public int[] applyOperations(int[] nums) {
        int n=nums.length;
        for(int i=0;i<n-1;i++){
            if(nums[i]==nums[i+1]){
                nums[i]*=2;
                nums[i+1]=0;
            }
        }
        int[] arr=new int[n];
        int i=0,j=n-1,k=0;
        while(i<n){
            if(nums[i]!=0){
                arr[k++]=nums[i];
                i++;
            }
            else{
                i++;
            }
        }
        return arr;
    }
}

File: 2551-apply-operations-to-an-array/2551-apply-operations-to-an-array.scala
object Solution {
    def applyOperations(nums: Array[Int]): Array[Int] = {
       val processed =  nums.indices.foldLeft(nums.clone()){
            (acc, i) =>
                acc.toIndexedSeq.lift(i+1) match {
                    case Some(next) if acc(i)==next && acc(i)!=0 =>
                        acc.updated(i, acc(i)*2).updated(i+1,0)
                    case _ => acc

                }

        }
        val nonZero = processed.filter(_ != 0)
        nonZero ++ Array.fill(nums.length - nonZero.length)(0)

        
    }
}